From da6ec2712e6fa8a2a5f42379175bfbc737be4199 Mon Sep 17 00:00:00 2001
From: rewine <lhongxu@outlook.com>
Date: Fri, 14 Apr 2023 14:38:20 +0800
Subject: [PATCH] feat: avoid use hardcode path

Log:
1. search for data files in XDG_DATA_DIRS

2. use cmake option to set path in dbus service
---
 CMakeLists.txt                                | 14 ++-----
 calendar-client/CMakeLists.txt                | 24 ++++--------
 ...service => com.deepin.Calendar.service.in} |  2 +-
 calendar-service/CMakeLists.txt               | 37 ++++++++-----------
 ...com.deepin.dataserver.Calendar.service.in} |  2 +-
 .../src/dbmanager/dhuanglidatabase.cpp        |  9 ++---
 calendar-service/src/main.cpp                 |  8 ++--
 .../src/synchronization/syncoperation.h       |  2 +-
 linglong.yaml                                 |  3 ++
 9 files changed, 41 insertions(+), 60 deletions(-)
 rename calendar-client/assets/dbus/{com.deepin.Calendar.service => com.deepin.Calendar.service.in} (60%)
 rename calendar-service/assets/data/{com.deepin.dataserver.Calendar.service => com.deepin.dataserver.Calendar.service.in} (51%)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 83fe1053..8d60a526 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -78,14 +78,8 @@ endforeach()
 
 qt5_create_translation(DTNG_QM_FILES
     ${DTNG_TS_FILES}
-     ${DTNG_QM_FILES}
-     )
-     
-if(DEFINED ENV{PREFIX})
-   set(CMAKE_INSTALL_PREFIX $ENV{PREFIX})
-else()
-   set(CMAKE_INSTALL_PREFIX /usr)
-endif()
+    ${DTNG_QM_FILES}
+)
   
- file(GLOB APP_QM_FILES "translations/*.qm")
- install(FILES ${APP_QM_FILES} DESTINATION ${CMAKE_INSTALL_DATADIR}/dde-calendar/translations)
+file(GLOB APP_QM_FILES "translations/*.qm")
+install(FILES ${APP_QM_FILES} DESTINATION ${CMAKE_INSTALL_DATADIR}/dde-calendar/translations)
diff --git a/calendar-client/CMakeLists.txt b/calendar-client/CMakeLists.txt
index 20efe13c..06a26c7a 100644
--- a/calendar-client/CMakeLists.txt
+++ b/calendar-client/CMakeLists.txt
@@ -1,8 +1,8 @@
 cmake_minimum_required(VERSION 3.7)
 
 if (NOT DEFINED VERSION)
-     message(WARNING "Not defined version ,about dialog version set 1.2.2")
-    set(VERSION 1.2.2)
+    message(WARNING "Not defined version, about dialog version set 5.10.0")
+    set(VERSION 5.10.0)
 endif ()
 
 #common resource names
@@ -10,7 +10,7 @@ set(APP_RES_DIR "assets")
 set(APP_BIN_NAME "dde-calendar")
 set(APP_DESKTOP "${APP_RES_DIR}/dde-calendar.desktop")
 set(APP_QRC "${APP_RES_DIR}/resources.qrc")
-set(APP_SERVICE "${APP_RES_DIR}/dbus/com.deepin.Calendar.service")
+set(APP_SERVICE_IN "${APP_RES_DIR}/dbus/com.deepin.Calendar.service.in")
 
 project(${APP_BIN_NAME})
 
@@ -19,7 +19,7 @@ set(CMAKE_INCLUDE_CURRENT_DIR ON)
 set(CMAKE_AUTOMOC ON)
 set(CMAKE_AUTORCC ON)
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wl,--as-needed -fPIE")
-set(CMAKE_EXE_LINKER_FLAGS "-pie")
+set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -pie")
 
 
 if (${CMAKE_SYSTEM_PROCESSOR} MATCHES "sw_64")
@@ -88,18 +88,10 @@ target_link_libraries(${PROJECT_NAME}
     commondata
 )
 
-#根据环境修改dbus服务文件
-if(DEFINED ENV{PREFIX})
-    set(APP_DBUS_SERVICE "[D-BUS Service]
-Name=com.deepin.Calendar
-Exec=/usr/bin/ll-cli run org.dde.calendar --exec dde-calendar")
-   file(WRITE ${APP_SERVICE} "${APP_DBUS_SERVICE}")
-
-   set(CMAKE_INSTALL_PREFIX $ENV{PREFIX})
-else()
-   set(CMAKE_INSTALL_PREFIX /usr)
-endif()
+#修改dbus服务文件
+set (EXEC_DDE_CALENDAR "${CMAKE_INSTALL_FULL_BINDIR}/${APP_BIN_NAME}" CACHE STRING "Command to execute dde-calendar")
 
+configure_file(${APP_SERVICE_IN} ${CMAKE_CURRENT_BINARY_DIR}/dbus/com.deepin.Calendar.service @ONLY)
 
 # Install files
 install(TARGETS dde-calendar DESTINATION ${CMAKE_INSTALL_BINDIR})
@@ -107,4 +99,4 @@ install(TARGETS dde-calendar DESTINATION ${CMAKE_INSTALL_BINDIR})
 install(DIRECTORY ${APP_RES_DIR}/dde-calendar
         DESTINATION ${CMAKE_INSTALL_DATADIR}/deepin-manual/manual-assets/application/)
 install(FILES ${APP_DESKTOP} DESTINATION ${CMAKE_INSTALL_DATADIR}/applications)
-install(FILES ${APP_SERVICE} DESTINATION ${CMAKE_INSTALL_DATADIR}/dbus-1/services)
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/dbus/com.deepin.Calendar.service DESTINATION ${CMAKE_INSTALL_DATADIR}/dbus-1/services)
diff --git a/calendar-client/assets/dbus/com.deepin.Calendar.service b/calendar-client/assets/dbus/com.deepin.Calendar.service.in
similarity index 60%
rename from calendar-client/assets/dbus/com.deepin.Calendar.service
rename to calendar-client/assets/dbus/com.deepin.Calendar.service.in
index 56be8dda..0e68070f 100644
--- a/calendar-client/assets/dbus/com.deepin.Calendar.service
+++ b/calendar-client/assets/dbus/com.deepin.Calendar.service.in
@@ -1,3 +1,3 @@
 [D-BUS Service]
 Name=com.deepin.Calendar
-Exec=/usr/bin/dde-calendar
+Exec=@EXEC_DDE_CALENDAR@
diff --git a/calendar-service/CMakeLists.txt b/calendar-service/CMakeLists.txt
index 79de333c..33a041e6 100644
--- a/calendar-service/CMakeLists.txt
+++ b/calendar-service/CMakeLists.txt
@@ -1,7 +1,6 @@
 cmake_minimum_required(VERSION 3.7)
 project(dde-calendar-service)
 
-
 set(CMAKE_CXX_STANDARD 11)
 set(CMAKE_INCLUDE_CURRENT_DIR ON)
 set(CMAKE_AUTOMOC ON)
@@ -10,7 +9,7 @@ set(CMAKE_AUTOMOC ON)
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fstack-protector-strong -D_FORTITY_SOURCE=1 -z noexecstack -pie -fPIC -z lazy")
 
 set(APP_RES_DIR "assets")
-set(APP_SERVICE "${APP_RES_DIR}/data/com.deepin.dataserver.Calendar.service")
+set(APP_SERVICE_IN "${APP_RES_DIR}/data/com.deepin.dataserver.Calendar.service.in")
 set(APP_SYSTEMD_SERVICE "${APP_RES_DIR}/data/com.dde.calendarserver.calendar.service")
 set(APP_SYSTEMD_TIMER "${APP_RES_DIR}/data/com.dde.calendarserver.calendar.timer")
 set(AUTOSTART_DESKTOP "${APP_RES_DIR}/dde-calendar-service.desktop")
@@ -51,7 +50,7 @@ QT5_ADD_RESOURCES(CALENDARSERVICE_SRCS ${APP_QRC})
 
 #后端程序自动退出宏控制
 if (NOT (CMAKE_BUILD_TYPE MATCHES Debug))
-add_definitions(-DCALENDAR_SERVICE_AUTO_EXIT)
+  add_definitions(-DCALENDAR_SERVICE_AUTO_EXIT)
 endif()
 
 add_executable(${PROJECT_NAME} ${CALENDARSERVICE_SRCS})
@@ -62,27 +61,23 @@ target_link_libraries(${PROJECT_NAME}
     ${LINK_LIBS}
     )
 
-#根据环境修改dbus服务文件
-if(DEFINED ENV{PREFIX})
-   set(APP_DBUS_SERVICE
-"[D-BUS Service]
-Name=com.deepin.dataserver.Calendar
-Exec=/usr/bin/ll-cli run  org.dde.calendar --exec  /opt/apps/org.dde.calendar/files/lib/deepin-daemon/dde-calendar-service
-")
-   message("${APP_DBUS_SERVICE}")
-   file(WRITE ${APP_SERVICE} "${APP_DBUS_SERVICE}")
-
-    set(CMAKE_INSTALL_PREFIX $ENV{PREFIX})
-else()
-    set(CMAKE_INSTALL_PREFIX /usr)
-endif()
+set (SERVICE_INSTALL_DIR "${CMAKE_INSTALL_FULL_LIBEXECDIR}/deepin-daemon" CACHE STRING "Install dir for dde-calendar-service")
+
+target_compile_definitions(${PROJECT_NAME}
+    PRIVATE
+    CALENDAR_SERVICE_PATH="${SERVICE_INSTALL_DIR}"
+)
+
+
+#修改dbus服务文件
+set (EXEC_DDE_CALENDAR_SERVICE "${SERVICE_INSTALL_DIR}/dde-calendar-service" CACHE STRING "Command to execute dde-calendar-service")
+
+configure_file(${APP_SERVICE_IN} ${CMAKE_CURRENT_BINARY_DIR}/dbus/com.deepin.dataserver.Calendar.service @ONLY)
 
-add_definitions(-DLINGLONG_PREFIX=\"${CMAKE_INSTALL_PREFIX}/\")
- 
 
 # Install files
-install(TARGETS ${PROJECT_NAME} DESTINATION lib/deepin-daemon/)
-install(FILES ${APP_SERVICE} DESTINATION ${CMAKE_INSTALL_DATADIR}/dbus-1/services/)
+install(TARGETS ${PROJECT_NAME} DESTINATION ${SERVICE_INSTALL_DIR})
+install(FILES ${CMAKE_CURRENT_BINARY_DIR}/dbus/com.deepin.dataserver.Calendar.service DESTINATION ${CMAKE_INSTALL_DATADIR}/dbus-1/services/)
 install(FILES ${AUTOSTART_DESKTOP} DESTINATION ${CMAKE_INSTALL_SYSCONFDIR}/xdg/autostart/)
 install(FILES ${HUANGLIDB} DESTINATION ${CMAKE_INSTALL_DATADIR}/dde-calendar/data/)
 install(FILES ${APP_SYSTEMD_SERVICE} DESTINATION lib/systemd/user/)
diff --git a/calendar-service/assets/data/com.deepin.dataserver.Calendar.service b/calendar-service/assets/data/com.deepin.dataserver.Calendar.service.in
similarity index 51%
rename from calendar-service/assets/data/com.deepin.dataserver.Calendar.service
rename to calendar-service/assets/data/com.deepin.dataserver.Calendar.service.in
index 2ce7991f..8336130c 100755
--- a/calendar-service/assets/data/com.deepin.dataserver.Calendar.service
+++ b/calendar-service/assets/data/com.deepin.dataserver.Calendar.service.in
@@ -1,4 +1,4 @@
 [D-BUS Service]
 Name=com.deepin.dataserver.Calendar
-Exec=/usr/lib/deepin-daemon/dde-calendar-service
+Exec=@EXEC_DDE_CALENDAR_SERVICE@
 
diff --git a/calendar-service/src/dbmanager/dhuanglidatabase.cpp b/calendar-service/src/dbmanager/dhuanglidatabase.cpp
index fdb764e1..67075d1e 100644
--- a/calendar-service/src/dbmanager/dhuanglidatabase.cpp
+++ b/calendar-service/src/dbmanager/dhuanglidatabase.cpp
@@ -9,14 +9,13 @@
 #include <QDebug>
 #include <QSqlError>
 
-#ifndef LINGLONG_PREFIX
-#define LINGLONG_PREFIX "/usr/"
-#endif
-
 DHuangLiDataBase::DHuangLiDataBase(QObject *parent)
     : DDataBase(parent)
 {
-    setDBPath(QString("%1share/dde-calendar/data/huangli.db").arg(LINGLONG_PREFIX));
+    QString huangliPath = QStandardPaths::locate(QStandardPaths::GenericDataLocation,
+                                                 QString("dde-calendar/data/huangli.db"),
+                                                 QStandardPaths::LocateFile);
+    setDBPath(huangliPath);
     setConnectionName("HuangLi");
     dbOpen();
 }
diff --git a/calendar-service/src/main.cpp b/calendar-service/src/main.cpp
index 20bfbe86..d40e91fb 100644
--- a/calendar-service/src/main.cpp
+++ b/calendar-service/src/main.cpp
@@ -12,15 +12,13 @@
 #include <QTranslator>
 #include <QCoreApplication>
 
-#ifndef LINGLONG_PREFIX
-#define LINGLONG_PREFIX "/usr/"
-#endif
-
-const static QString CalendarServiceTranslationsDir = QString("%1share/dde-calendar/translations").arg(LINGLONG_PREFIX);
 
 bool loadTranslator(QCoreApplication *app, QList<QLocale> localeFallback = QList<QLocale>() << QLocale::system())
 {
     bool bsuccess = false;
+    QString CalendarServiceTranslationsDir = QStandardPaths::locate(QStandardPaths::GenericDataLocation,
+                                                  		    QString("dde-calendar/translations"),
+								    QStandardPaths::LocateDirectory);
     for (auto &locale : localeFallback) {
         QString translateFilename = QString("%1_%2").arg(app->applicationName()).arg(locale.name());
         QString translatePath = QString("%1/%2.qm").arg(CalendarServiceTranslationsDir).arg(translateFilename);
diff --git a/calendar-service/src/synchronization/syncoperation.h b/calendar-service/src/synchronization/syncoperation.h
index fb954352..b20cae9f 100644
--- a/calendar-service/src/synchronization/syncoperation.h
+++ b/calendar-service/src/synchronization/syncoperation.h
@@ -31,7 +31,7 @@
 #define SYNC_Invalid_File_Size          7510            /*文件大小不合法*/
 #define SYNC_Metadata_Check_Error       7511            /*元数据校验出错*/
 
-const QString utcloudcalendatpath = "/usr/lib/deepin-daemon/dde-calendar-service";
+const QString utcloudcalendatpath = CALENDAR_SERVICE_PATH"/dde-calendar-service";
 
 using SyncInter = com::deepin::sync::cloudopt;
 
diff --git a/linglong.yaml b/linglong.yaml
index 170d284b..48c25f96 100644
--- a/linglong.yaml
+++ b/linglong.yaml
@@ -33,3 +33,6 @@ source:
 
 build:
   kind: cmake
+  extra_args: |
+    -DEXEC_DDE_CALENDAR="/usr/bin/ll-cli run org.dde.calendar --exec dde-calendar"
+    -DEXEC_DDE_CALENDAR_SERVICE="/usr/bin/ll-cli run org.dde.calendar --exec /opt/apps/org.dde.calendar/files/lib/deepin-daemon/dde-calendar-service"

